pytradfri/resource.py

```diff
@@ -1,49 +1,71 @@
 """Resources for devices."""
+from __future__ import annotations
+
 from datetime import datetime
+from typing import Union, cast, Callable, Any, Dict, List
+from abc import abstractmethod
+
+from .command import Command, TYPE_PROCESS_RESULT_CB, TYPE_ERR_CB
+from .const import (
+    ATTR_NAME,
+    ATTR_CREATED_AT,
+    ATTR_ID,
+)
 
-from .command import Command
-from .const import ATTR_CREATED_AT, ATTR_ID, ATTR_NAME
+# type alias
+TYPE_RAW = Dict[str, Union[str, int, List[Dict[str, Union[str, int]]]]]
+TYPE_RAW_SIMPLE = Dict[str, Union[str, int]]
+TYPE_RAW_LIST = Dict[str, List[Dict[str, Union[str, int]]]]
 
 
 class ApiResource:
     """Base object for resources returned from the gateway."""
 
-    def __init__(self, raw):
+    def __init__(self, raw: TYPE_RAW) -> None:
         """Initialize base object."""
         self.raw = raw
 
     @property
-    def id(self):
+    def id(self) -> str:
         """Id."""
-        return self.raw.get(ATTR_ID)
+        return str(self.raw.get(ATTR_ID))
```

Comment:
We don't want to convert `None` to a string if the key is missing in the dict.

Reply:
Corrected. Remark the cast is needed because mypy claims self.raw.get return a RAW_TYPE.